README.md
# Introduction
The managed cloud provider is a provider extension to the `DockerParallel` package. It aims to provide container management tools for the non-kubernetes cloud provider. The cloud provider only needs to define functions to run and stop the worker. The managed cloud provider can automatically scale the number of workers up or down according to the user's request. Below shows the structure of the managed cloud provider

![](vignettes/Inheritance.jpg)

where `CloudProvider` is the base class defined in `DockerParallel` package. `CloudProvider` dose nothing but define a few generics(the virtual class in C++ language)  .`ManagedCloudProvider` inherits from `CloudProvider`, it takes over the control of the worker container by defining the method for `setDockerWorkerNumber` and `getDockerWorkerNumbers`(The red font). The provider who inherits from `ManagedCloudProvider` does not need to manage the workers itself, it only needs to define the method for the generics exposed by `ManagedCloudProvider`, namely

1. runDockerWorkerContainers
2. getDockerWorkerStatus
3. IsDockerWorkerInitializing
4. IsDockerWorkerRunning
5. IsDockerWorkerStopped
6. killDockerWorkerContainers

The name should be self-explained. Please note that only `runDockerWorkerContainers`, `getDockerWorkerStatus` and `killDockerWorkerContainers` are required for making the provider works. The rest are optional and have the default methods(based on `getDockerWorkerStatus`). 

# runDockerWorkerContainers
`runDockerWorkerContainers` should deploy the worker on the cloud. The generic for `runDockerWorkerContainers` is 

```r
getGeneric("runDockerWorkerContainers")
#> nonstandardGenericFunction for "runDockerWorkerContainers" defined from package "ManagedCloudProvider"
#> 
#> function (provider, cluster, container, hardware, workerNumber, 
#>     verbose) 
#> {
#>     standardGeneric("runDockerWorkerContainers")
#> }
#> <bytecode: 0x000002161f812670>
#> <environment: 0x000002161f943478>
#> Methods may be defined for arguments: provider
#> Use  showMethods(runDockerWorkerContainers)  for currently available ones.
```
where `provider` is the cloud provider object and `cluster` is the `DockerCluster` object, `container` specifies which container should be used for the worker, `hardware` indicates the hardware requirement for each worker, `workerNumber` is the number of the workers that should be deployed. `verbose` is the verbose level. The return value of the function should be a character vector of length `workerNumber` with each element corresponding to a handle of a worker. The handle should be used to manage the lifecycle of the workers. It is possible to return the same handle for multiple workers if they share the same lifecycle(e.g. multiple workers share the same container). 
    
# getDockerWorkerStatus
`getDockerWorkerStatus` is for checking the status of the workers, a worker can have three possible status, namely `initializing`, `running` and `stopped`. The generic for `getDockerWorkerStatus` is

```r
getGeneric("getDockerWorkerStatus")
#> nonstandardGenericFunction for "getDockerWorkerStatus" defined from package "ManagedCloudProvider"
#> 
#> function (provider, cluster, workerHandles, verbose) 
#> {
#>     standardGeneric("getDockerWorkerStatus")
#> }
#> <bytecode: 0x0000021629794980>
#> <environment: 0x0000021629777048>
#> Methods may be defined for arguments: provider
#> Use  showMethods(getDockerWorkerStatus)  for currently available ones.
```
where `workerHandles` is a vector of handles returned by `runDockerWorkerContainers`, the elements of `workerHandles` are unique. The return value should be a character vector indicating the status of each worker.

The generics `IsDockerWorkerInitializing`, `IsDockerWorkerRunning` and `IsDockerWorkerStopped` check whether the worker is in a specific status. Once you have `getDockerWorkerStatus` defined these methods are automatically available. Therefore, you do not have to define them unless you have any faster implementation.

# killDockerWorkerContainers
`killDockerWorkerContainers` will kill the workers given the worker handles. If a handle is shared by multiple workers, all workers with the same handle should be killed. The generic is

```r
getGeneric("killDockerWorkerContainers")
#> nonstandardGenericFunction for "killDockerWorkerContainers" defined from package "ManagedCloudProvider"
#> 
#> function (provider, cluster, workerHandles, verbose) 
#> {
#>     standardGeneric("killDockerWorkerContainers")
#> }
#> <bytecode: 0x000002161fc45950>
#> <environment: 0x000002161fc7bd10>
#> Methods may be defined for arguments: provider
#> Use  showMethods(killDockerWorkerContainers)  for currently available ones.
```
The return value is a logical vector indicating whether the killing process is successful. Please note that the method should accept the invalid handle. That is, if the worker has already been killed or the handle does not exist, the method should not raise any error and the killing action should be denoted as successful. 






NEWS
                          Changes in version 1.0.0
First public release.




tests/testthat.R
library(testthat)
library(ManagedCloudProvider)

test_check("ManagedCloudProvider")


tests/testthat/test_generalTest.R
## Test the general unit test function exported from this package

resetDummyManagedProvider()
verbose <- 0

test_that("testthat", {
    expect_error(
        provider <- DummyManagedProvider()
        ,NA)
    expect_error(
        container <- DummyWorkerContainer()
        ,NA)
    expect_error(
        generalDockerClusterTest(
            cloudProvider = provider,
            workerContainer = container,
            workerNumber = 5,
            testReconnect = TRUE,
            verbose = verbose)
        ,NA)
})


tests/testthat/test_ManagedProvider.R
resetDummyManagedProvider()
cluster <- NULL
verbose <- 0L

test_that("DockerCluster constructor", {
    expect_error(
        provider <- DummyManagedProvider()
        ,NA)
    expect_error(
        container <- DummyWorkerContainer()
        ,NA)
    expect_error(
        cluster <<- makeDockerCluster(
            cloudProvider = provider,
            workerContainer = container,
            verbose = verbose)
        ,NA)
})

test_that("DockerCluster print", {
    expect_error(
        capture.output(show(cluster)),
        NA
    )
})

test_that("DockerCluster set worker number", {
    expect_error(
        cluster$setWorkerNumber(10),
        NA
    )
    expect_equal(.getExpectedWorkerNumber(cluster), 10)
    expect_identical(cluster$getWorkerNumbers(),
                     list(initializing = 0L, running = 0L, expected = 10L))

})

test_that("DockerCluster run server", {
    expect_error(
        cluster$startServer(),
        NA
    )
    expect_true(cluster@cloudProvider$initialized)
    expect_true(cluster$isServerRunning())
    expect_identical(cluster$getWorkerNumbers(),
                     list(initializing = 0L, running = 0L, expected = 10L))
    expect_equal(cluster@cloudRuntime$serverPublicIp, "8.8.8.8")
    expect_equal(cluster@cloudRuntime$serverPublicPort, 123)
    expect_equal(cluster@cloudRuntime$serverPrivateIp, "192.168.1.1")
    expect_equal(cluster@cloudRuntime$serverPrivatePort, 456)

    expect_identical(cluster@cloudProvider$serverContainer$image, "serverImage")
})


test_that("DockerCluster run worker", {
    expect_error(
        cluster$setWorkerNumber(5),
        NA
    )
    expect_identical(cluster$getWorkerNumbers(),
                     list(initializing = 0L, running = 5L, expected = 5L))
    expect_identical(cluster@cloudProvider$workerContainer$image, "workerImage")
})

test_that("DockerCluster update", {
    expect_error(
        cluster$update(),
        NA
    )
})


test_that("DockerCluster register backend", {
    expect_error(
        cluster$registerBackend(),
        NA
    )
    expect_error(
        cluster$deregisterBackend(),
        NA
    )
})

test_that("DockerCluster stop cluster", {
    expect_error(
        cluster$stopCluster(),
        NA
    )
    expect_identical(cluster$getWorkerNumbers(),
                     list(initializing = 0L, running = 0L, expected = 5L))
    expect_false(cluster$isServerRunning())

    expect_error(
        cluster$update(),
        NA
    )
})


test_that("DockerCluster reconnect: 1 server 5 worker", {
    gc()
    provider1 <- DummyManagedProvider()
    container1 <- DummyWorkerContainer()
    cluster1 <- makeDockerCluster(
        cloudProvider = provider1,
        workerContainer = container1,
        workerNumber = 5,
        verbose = verbose)
    expect_error(
        cluster1$startCluster(),
        NA
    )
    expect_identical(
        cluster1$getWorkerNumbers(),
        list(initializing = 0L, running = 5L, expected = 5L)
    )

    provider2 <- DummyManagedProvider()
    container2 <- DummyWorkerContainer()
    cluster2 <- makeDockerCluster(
        cloudProvider = provider2,
        workerContainer = container2,
        workerNumber = 0,
        verbose = verbose)
    expect_error(cluster2$reconnect(),
                 NA)
    expect_identical(
        cluster1$getWorkerNumbers(),
        list(initializing = 0L, running = 5L, expected = 5L)
    )
})

test_that("DockerCluster reconnect: 1 server no worker", {
    gc()
    provider1 <- DummyManagedProvider()
    container1 <- DummyWorkerContainer()
    cluster1 <- makeDockerCluster(
        cloudProvider = provider1,
        workerContainer = container1,
        workerNumber = 0,
        verbose = verbose)
    expect_error(
        cluster1$startCluster(),
        NA
    )
    expect_identical(
        cluster1$getWorkerNumbers(),
        list(initializing = 0L, running = 0L, expected = 0L)
    )

    provider2 <- DummyManagedProvider()
    container2 <- DummyWorkerContainer()
    cluster2 <- makeDockerCluster(
        cloudProvider = provider2,
        workerContainer = container2,
        workerNumber = 0,
        verbose = verbose)
    expect_error(cluster2$reconnect(),
                 NA)
    expect_identical(
        cluster1$getWorkerNumbers(),
        list(initializing = 0L, running = 0L, expected = 0L)
    )
})

test_that("DockerCluster auto remove the cluster", {
    gc()
    provider <- DummyManagedProvider()
    container <- DummyWorkerContainer()
    cluster <- makeDockerCluster(
        cloudProvider = provider,
        workerContainer = container,
        verbose = verbose)
    expect_error(cluster$reconnect())
})



R/utils.R
removeWorkersHandle <- function(provider, index){
    if(length(index)!=0){
        provider$workerHandles <- provider$workerHandles[-index]
        provider$workerPerHandle <- provider$workerPerHandle[-index]
    }
}

removeDiedWorkers <- function(cluster){
    verbose <- cluster$verbose
    provider <- .getCloudProvider(cluster)
    workerStatus <- NULL
    if(length(provider$workerHandles)!=0){
        workerStatus <- getDockerWorkerStatus(provider, cluster, provider$workerHandles, verbose=verbose)
        if(any(workerStatus=="stopped")){
            removeWorkersHandle(provider, which(workerStatus=="stopped"))
            workerStatus <- workerStatus[workerStatus != "stopped"]
        }
    }
    workerStatus
}

getManagedWorkerNumber <- function(provider){
    sum(provider$workerPerHandle)
}



addManagedWorkersInternal <- function(cluster, container, hardware, workerNumber){
    provider <- .getCloudProvider(cluster)
    workerHandles <- runDockerWorkerContainers(provider = provider,
                                               cluster = cluster,
                                               container = container,
                                               hardware = hardware,
                                               workerNumber = workerNumber,
                                               verbose = cluster$verbose)
    addManagedWorkerHandles(provider, workerHandles)
    invisible(NULL)
}


myknapsack <- function (workerPerHandle, killedWorkerNum)
{
    idx <- which(workerPerHandle<=killedWorkerNum)
    if(length(idx)==0){
        return(list(capacity=0, indices = c()))
    }
    KnapsackSolution <-
        adagio::knapsack(workerPerHandle[idx],
                         workerPerHandle[idx],
                         killedWorkerNum)
    KnapsackSolution$indices <- idx[KnapsackSolution$indices]
    KnapsackSolution
}


removeManagedWorkersInternal <- function(cluster, workerNumber){
    verbose <- cluster$verbose
    provider <- .getCloudProvider(cluster)

    ## Find which workers will be killed while satisfying
    ## that the killed workers is less than or equal to workerNumber
    KnapsackSolution <-
        myknapsack(provider$workerPerHandle,
                   workerNumber)
    killedWorkerNumber <- KnapsackSolution$capacity
    killedInstanceIndex <- KnapsackSolution$indices
    if(killedWorkerNumber < workerNumber){
        if(killedWorkerNumber==0){
            message("No worker can be killed as all containers have more than ",
                    workerNumber,
                    " workers")
        }else{
            message("Only ", killedWorkerNumber,
                    " workers will be killed as multiple workers share the same container")
        }
    }
    if(killedWorkerNumber!=0){
        success <- killDockerWorkerContainers(
            provider,
            cluster,
            workerHandles = provider$workerHandles[killedInstanceIndex],
            verbose = verbose)
        if(any(!success)){
            warning("Fail to kill some worker containers")
        }
        removeWorkersHandle(provider, killedInstanceIndex[success])
    }
    invisible(NULL)
}

verbosePrint<-function(verbose, ...){
    if(verbose)
        message(...)
}


R/export.R
#' Add or get the worker container handles to the managed cloud provider
#'
#' Add or get the worker container handles to the managed cloud provider. The
#' handles can be duplicated if multiple workers share the same container
#'
#' @param provider A `ManagedCloudProvider` object
#' @param handles the worker container handles
#' @rdname managedHandles
#' @examples
#' ## make a dummy provider
#' DummyProvider <- setRefClass("DummyProvider", contains = "ManagedCloudProvider")
#' provider <- DummyProvider()
#'
#' ## No worker handle in the provider
#' getManagedWorkerHandles(provider)
#'
#' ## Add worker handles
#' addManagedWorkerHandles(provider, c("a", "b"))
#' getManagedWorkerHandles(provider)
#'
#' ## It is possible to add the same handle
#' ## if multiple workers share the same container
#' addManagedWorkerHandles(provider, c("a"))
#' getManagedWorkerHandles(provider)
#' @returns
#' addManagedWorkerHandles: No return value
#' getManagedWorkerHandles: A character vector of the worker handles
#' @export
addManagedWorkerHandles <- function(provider, handles){
    if(length(handles)!=0){
        allHandles <- c(getManagedWorkerHandles(provider), handles)
        info <- table(allHandles)
        provider$workerHandles <- names(info)
        provider$workerPerHandle <- as.vector(info)
        invisible(NULL)
    }
}


#' @rdname managedHandles
#' @export
getManagedWorkerHandles <- function(provider){
    workerHandles <- provider$workerHandles
    workerPerHandle <- provider$workerPerHandle
    x <- lapply(seq_along(workerHandles),
                function(i)rep(workerHandles[i],workerPerHandle[i])
    )
    unlist(x)
}


R/adaptors.R
#' Set the worker number in the cluster
#'
#' The developer should define `runDockerWorkerContainers`, `getDockerWorkerStatus`
#' and `killDockerWorkerContainers` instead.
#'
#' @inheritParams  DockerParallel::setDockerWorkerNumber
#' @param workerNumber Integer(1), the worker number to be set
#' @return No return value
#' @export
setMethod("setDockerWorkerNumber", "ManagedCloudProvider",
          function(provider, cluster, container, hardware, workerNumber, verbose)
          {
              verbosePrint(verbose, "Setting worker number to ", workerNumber)
              if(cluster$isServerRunning()){
                  ## set the expected worker number
                  removeDiedWorkers(cluster)

                  workerOffset <- workerNumber - getManagedWorkerNumber(provider)
                  if(workerOffset > 0){
                      addManagedWorkersInternal(cluster = cluster,
                                                container = container,
                                                hardware = hardware,
                                                workerNumber = workerOffset)
                  }
                  if(workerOffset < 0){
                      removeManagedWorkersInternal(cluster = cluster,
                                                   workerNumber = abs(workerOffset))
                  }
              }
          })

#' get the worker number in the cluster
#'
#' The developer should define `runDockerWorkerContainers`, `getDockerWorkerStatus`
#' and `killDockerWorkerContainers` instead.
#'
#' @inheritParams  DockerParallel::getDockerWorkerNumbers
#' @return A list with `initializing` and `running` integer elements
#' @export
setMethod("getDockerWorkerNumbers", "ManagedCloudProvider", function(provider, cluster, verbose){
    if(cluster$isServerRunning()){
        workerStatus <- removeDiedWorkers(cluster)
        list(initializing = sum(provider$workerPerHandle[workerStatus=="initializing"]),
             running = sum(provider$workerPerHandle[workerStatus=="running"])
             )
    }else{
        list(initializing = 0L, running = 0L)
    }
})





R/all-generics.R
##################################################
##     Worker management method 2:
##    DockerCluster controls the lifecycle of the workers
##################################################
#' Run the worker container
#'
#' Run the workers and return a character vector of the worker handles. Each handle must
#' correspond to a container. The handle can be duplicated if multiple workers share the
#' same container. There is no default method for this generic.
#'
#'
#' @inheritParams DockerParallel::initializeCloudProvider
#' @param workerNumber Integer. The number of workers needs to be run.
#' @param container S4 `DockerContainer` Object. The worker container.
#' @param hardware S4 `DockerHardware` Object. The worker hardware.
#'
#'
#' @returns
#' A character vector with each element corresponding to a worker container.
#' The length must be equal to `workerNumber`
#' @export
setGeneric("runDockerWorkerContainers",
           function(provider, cluster, container, hardware, workerNumber, verbose){
               standardGeneric("runDockerWorkerContainers")
           },
           signature = "provider")


#' Get the worker status
#'
#' Get the worker status. Unless you have a faster implementation, you only need to
#' define `getDockerWorkerStatus`. The function should return a character vector with
#' each element corresponding to a worker in `workerHandles`.
#' Each element must be one of three possible characters `"initializing"`, `"running"` or
#' `"stopped"`. There is no default method for `getDockerWorkerStatus`.
#'
#'
#' @inheritParams DockerParallel::initializeCloudProvider
#' @param workerHandles Character(n). A character vector of **unique** instance handles.
#'
#' @rdname workerStatus
#' @returns
#' `getDockerWorkerStatus` : A character vector with each element corresponding
#' to an instance in `workerHandles`. Each element must be one of three possible characters
#' `"initializing"`, `"running"` or `"stopped"`
#'
#' `IsDockerWorkerInitializing`, `IsDockerWorkerRunning`, `IsDockerWorkerStopped`:
#' A logical vector with each element corresponding to the status of each instance
#' @export
setGeneric("getDockerWorkerStatus", function(provider, cluster, workerHandles, verbose){
    standardGeneric("getDockerWorkerStatus")
},
signature = "provider")


#' @rdname workerStatus
#' @export
setGeneric("IsDockerWorkerInitializing", function(provider, cluster, workerHandles, verbose){
    standardGeneric("IsDockerWorkerInitializing")
},
signature = "provider")


#' @rdname workerStatus
#' @export
setGeneric("IsDockerWorkerRunning", function(provider, cluster, workerHandles, verbose){
    standardGeneric("IsDockerWorkerRunning")
},
signature = "provider")


#' @rdname workerStatus
#' @export
setGeneric("IsDockerWorkerStopped", function(provider, cluster, workerHandles, verbose){
    standardGeneric("IsDockerWorkerStopped")
},
signature = "provider")

#' Kill the worker container
#'
#' Kill the worker container. The worker handles are unique.
#' If multiple workers share the same instance, all workers in the
#' same container should be killed. There is no default method for this generic.
#'
#' @inheritParams getDockerWorkerStatus
#'
#' @returns
#' A logical vector indicating whether the killing operation is successful for each instance
#' @export
setGeneric("killDockerWorkerContainers", function(provider, cluster, workerHandles, verbose){
    standardGeneric("killDockerWorkerContainers")
},
signature = "provider")


R/dummyManagedProvider.R
handlePrefix <- "DummyManagedProviderHandle_"
getHandleName <- function(i){
    paste0(handlePrefix, i)
}
getAllHandles <- function(){
    envs <- Sys.getenv()
    idx <- startsWith(names(envs), handlePrefix)
    handles <- envs[idx]
    handleNames <- gsub(handlePrefix, "", names(handles), fixed = TRUE)
    names(handles) <- handleNames
    for(i in seq_along(handles)){
        handles[i] <- as.integer(handles[i])
    }
    handles
}

resetDummyManagedProvider <- function(){
    Sys.setenv(DummyManagedProvider = "")
    Sys.setenv(DummyManagedProviderClusterData = "")
    envs <- Sys.getenv()
    idx <- startsWith(names(envs), handlePrefix)
    handles <- envs[idx]
    workerNames <- names(handles)
    for(i in seq_along(workerNames)){
        arg <- list("")
        names(arg) <- i
        do.call(Sys.setenv, arg)
    }
    invisible()
}
#########################################################
.DummyManagedProvider <- setRefClass("DummyManagedProvider",
                              fields =
                                  list(
                                      initialized = "logical",
                                      isServerRunning = "logical",
                                      cleanup = "logical",
                                      handleIdx = "integer",
                                      serverContainer = "DockerContainer",
                                      workerContainer = "DockerContainer"),
                              contains = "ManagedCloudProvider")


DummyManagedProvider <- function(initialized = FALSE,
                          isServerRunning = FALSE,
                          cleanup = FALSE){
    .DummyManagedProvider(initialized = initialized,
                   isServerRunning = isServerRunning,
                   cleanup = cleanup,
                   workerHandles = character(0),
                   handleIdx = 0L)
}



setMethod("initializeCloudProvider", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(!provider$initialized)

    provider$initialized <- TRUE
})


setMethod("runDockerServer", "DummyManagedProvider", function(provider, cluster, container, hardware, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(container, "DockerContainer"))
    stopifnot(is(hardware, "DockerHardware"))
    stopifnot(is(verbose, "numeric"))

    stopifnot(!provider$isServerRunning)
    provider$isServerRunning <- TRUE
    provider$cleanup <- FALSE
    provider$serverContainer <- container
    Sys.setenv(DummyManagedProvider = .getJobQueueName(cluster))

    clusterData <- serialize(getDockerStaticData(cluster), NULL)
    encodedValue <- jsonlite::base64_enc(clusterData)
    Sys.setenv(DummyManagedProviderClusterData = encodedValue)
})

setMethod("stopDockerServer", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))

    stopifnot(provider$isServerRunning)
    provider$isServerRunning <- FALSE
    Sys.setenv(DummyManagedProvider = "")
    Sys.setenv(DummyManagedProviderClusterData = "")
})

setMethod("getServerStatus", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))

    if(provider$isServerRunning){
        return("running")
    }else{
        return("stopped")
    }
})

setMethod("getDockerServerIp", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(provider$isServerRunning)

    list(publicIp = "8.8.8.8",
         publicPort = 123,
         privateIp = "192.168.1.1",
         privatePort = 456)
})


#' @describeIn runDockerWorkerContainers The method for the dummy managed provider
setMethod("runDockerWorkerContainers", "DummyManagedProvider", function(provider, cluster, container, hardware, workerNumber, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(container, "DockerContainer"))
    stopifnot(is(hardware, "DockerHardware"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(cluster$isServerRunning())

    workerPerContainer <- 4
    handleIdxBegin <- provider$handleIdx
    handleIdxEnd <- handleIdxBegin + ceiling(workerNumber/workerPerContainer) - 1
    provider$handleIdx <- as.integer(handleIdxEnd + 1)
    handles <- unlist(lapply(handleIdxBegin:handleIdxEnd, function(x) as.character(rep(x, workerPerContainer))))
    handles <- handles[seq_len(workerNumber)]
    provider$workerContainer <- container

    workerInContainer <- table(handles)
    for(i in seq_along(workerInContainer)){
        name <- getHandleName(names(workerInContainer)[i])
        value <- workerInContainer[i]
        arg <- list(value)
        names(arg) <- name
        do.call(Sys.setenv, arg)
    }
    handles
})

#' @describeIn workerStatus The method for the dummy managed provider
setMethod("getDockerWorkerStatus", "DummyManagedProvider", function(provider, cluster, workerHandles, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))

    allHandles <- getAllHandles()
    idx <- workerHandles%in% names(allHandles)
    status <- rep("running", length(workerHandles))
    status[!idx] <- "stopped"
    status
})

#' @describeIn killDockerWorkerContainers The method for the dummy managed provider
setMethod("killDockerWorkerContainers", "DummyManagedProvider", function(provider, cluster, workerHandles, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(length(workerHandles) == length(unique(workerHandles)))

    for(i in workerHandles){
        handleName <- getHandleName(i)
        Sys.unsetenv(handleName)
    }
    rep(TRUE, length(workerHandles))
})


setMethod("dockerClusterExists", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))

    Sys.getenv("DummyManagedProvider") == .getJobQueueName(cluster)
})

setMethod("reconnectDockerCluster", "DummyManagedProvider", function(provider, cluster, verbose){
    stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(!provider$isServerRunning)
    stopifnot(Sys.getenv("DummyManagedProvider") == .getJobQueueName(cluster))
    encodedValue <- Sys.getenv("DummyManagedProviderClusterData")
    staticData <- unserialize(jsonlite::base64_dec(encodedValue))
    setDockerStaticData(cluster, staticData)
    allHandles <- getAllHandles()
    addManagedWorkerHandles(provider, names(allHandles))
    provider$isServerRunning <- TRUE
})


setMethod("cleanupDockerCluster", "DummyManagedProvider", function(provider, cluster, verbose){
    # stopifnot(provider$initialized)
    stopifnot(is(cluster, "DockerCluster"))
    stopifnot(is(verbose, "numeric"))
    stopifnot(!provider$isServerRunning)
    stopifnot(!provider$cleanup)
    stopifnot(Sys.getenv("DummyManagedProvider") == "")
    stopifnot(length(getAllHandles()) == 0)

    provider$cleanup <- TRUE

})
